scheds/rust/scx_p2dq/src/main.rs

```diff
@@ -0,0 +1,343 @@
+        // Open the BPF prog first for verification.
+        let mut skel_builder = BpfSkelBuilder::default();
+        skel_builder.obj_builder.debug(opts.verbose > 1);
+        init_libbpf_logging(None);
+        info!(
+            "Running scx_p2dq (build ID: {})",
+            build_id::full_version(env!("CARGO_PKG_VERSION"))
+        );
+        let mut skel = scx_ops_open!(skel_builder, open_object, p2dq).unwrap();
+
+        let topo = Topology::new()?;
+        if opts.init_dsq_index > opts.dumb_queues - 1 {
+            panic!("Invalid init_dsq_index {}", opts.init_dsq_index);
+        }
+        if opts.dsq_time_slices.len() > 0 {
+            if opts.dsq_time_slices.len() != opts.dumb_queues {
+                panic!(
+                    "Invalid number of dsq_time_slices, got {} need {}",
+                    opts.dsq_time_slices.len(),
+                    opts.dumb_queues,
+                )
+            }
+            for vals in opts.dsq_time_slices.windows(2) {
+                assert!(
+                    vals[0] < vals[1],
+                    "DSQ time slices must be in increasing order"
+                );
+            }
+            for (i, slice) in opts.dsq_time_slices.iter().enumerate() {
+                skel.maps.bss_data.dsq_time_slices[i] = slice * 1000;
+            }
+        } else {
+            for i in 0..=opts.dumb_queues {
+                skel.maps.bss_data.dsq_time_slices[i] =
+                    dsq_slice_ns(i as u64, opts.min_slice_us, opts.dsq_shift);
+            }
+        }
+
+        if opts.partial {
+            skel.struct_ops.p2dq_mut().flags |= *compat::SCX_OPS_SWITCH_PARTIAL;
+        }
+
+        skel.maps.rodata_data.min_slice_us = opts.min_slice_us;
+        skel.maps.rodata_data.dsq_shift = opts.dsq_shift as u64;
+        skel.maps.rodata_data.kthreads_local = !opts.disable_kthreads_local;
+        skel.maps.rodata_data.debug = opts.verbose as u32;
+        skel.maps.rodata_data.nr_cpus = *NR_CPU_IDS as u32;
+        skel.maps.rodata_data.nr_dsqs_per_llc = opts.dumb_queues as u32;
+        skel.maps.rodata_data.init_dsq_index = opts.init_dsq_index as i32;
+        skel.maps.rodata_data.nr_llcs = topo.all_llcs.clone().keys().len() as u32;
+        skel.maps.rodata_data.nr_nodes = topo.nodes.clone().keys().len() as u32;
+        skel.maps.rodata_data.has_little_cores = topo.has_little_cores();
+        skel.maps.rodata_data.smt_enabled =
+            read_file_usize(Path::new("/sys/devices/system/cpu/smt/active")).unwrap_or(0) == 1;
+
+        let mut skel = scx_ops_load!(skel, p2dq, uei)?;
+
+        let mut cpu_ctxs: Vec<bpf_intf::cpu_ctx> = vec![];
+        let cpu_ctxs_vec = skel
+            .maps
+            .cpu_ctxs
+            .lookup_percpu(&0u32.to_ne_bytes(), libbpf_rs::MapFlags::ANY)?
+            .unwrap();
+
+        for (cpu_id, cpu_arc) in topo.all_cpus {
+            let cpu = cpu_arc.clone();
```

Comment:
nit: do we need this clone/variable? It's cloning the `Arc<Cpu>`, incrementing the reference count, doing read only things, then dropping it at the end. These operations should all be possible on the `&Arc<Cpu> cpu_arc` (which could be renamed to `cpu`).